![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**


**DISCLAIMER 📣**

**En este proyecto utilizamos un enfoque de top-N recomendaciones, ya que observamos que el modelo tiende a asignar probabilidades bajas y muy cercanas entre sí debido al fuerte desbalance de clases. Bajo este escenario, elegir un umbral fijo generaba poca estabilidad y un rendimiento inconsistente entre batches. En cambio, seleccionar siempre los N productos más probables por cliente permitió controlar mejor el volumen de recomendaciones, reducir la sensibilidad al threshold y obtener resultados más coherentes con el comportamiento real del modelo. NO ES NECESARIO QUE HAGAN ESTE MISMO PROCEDIMIENTO, ES SOLO PARA GUIARLOS**

**El desempeño de su modelo dependerá de como crearon los datos históricos de transacciones por lo que les puede tomar mucho tiempo o poco el entrenamiento, les adjuntamos un codigo de como lo trabajamos en la sección de análisis. Sus análisis pueden variar en las distintas pruebas de los batches, no se desanimen si les dan malos resultados. Procuren de realizar un análisis por semana en conjunto con un análisis general de todos los batches.**

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - **Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)**

<center>
<img src="https://media1.giphy.com/media/ZkIUY6LZJMwrSMQMWu/giphy.gif?cid=6c09b952msc755426bpsadn4sysofcpqczczi7x1pghd41sw&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=v" width="400" height="300">

# 📬 Entrega Parcial 3 (30% del Proyecto)

## 📖 Enunciado

Después de haber superado con éxito la implementación de sus flujos de trabajo con `Airflow`, el equipo **Deep Drinkers 🤖** ha demostrado que no solo sabe construir modelos, sino también integrarlos en pipelines reales listos para producción.

El equipo técnico de **SodAI Drinks 🥤** ha quedado tan satisfecho con la automatización de sus procesos que ahora les ha encomendado una nueva, y final, misión:  
- 📊 **Evaluar la calidad real de sus modelos en un entorno más desafiante** 
- 🧠 **Comunicar los hallazgos de forma clara y profesional frente a una comisión**.

<center>
<img src="https://i.giphy.com/btbUGSHh3f6eBjbDfh.webp" width="350">


## 📌 Predicciones
<center>
<img src="https://i.giphy.com/NTMxntb8rMzmbd1x97.webp" width="500" height="300">
</center>

Su *pipeline* automatizado debe demostrar que puede manejar datos reales en producción. **SodAI Drinks** 🥤 liberará *batches* de datos en fechas diferentes para evaluar si su sistema se adapta correctamente, los cuales simularán la llegada de información nueva (cada batch es una semana).

Lo que debe funcionar automáticamente:
- **Ejecución Automatizada**: Su DAG de `Airflow` debe ser capaz de procesar cada *batch* de datos de manera automática cuando esté disponible.
- **Monitoreo Continuo**: El *pipeline* debe evaluar la calidad de los datos entrantes y detectar posibles desviaciones respecto al conjunto original.
- **Decisión de Reentrenamiento**: Basándose en los análisis de *data drift* y métricas de rendimiento, el sistema debe determinar automáticamente si es necesario reentrenar el modelo.
- **Tracking Completo**: Todas las ejecuciones, métricas y decisiones deben quedar registradas en `MLFlow` para posterior análisis.
- **Generación de Predicciones**: Para cada *batch* procesado, el sistema debe generar las predicciones correspondientes y almacenarlas de manera organizada.

**IMPORTANTE: Es imperativo que guarden los resultados de su pipeline (métricas, tracking, predicciones, etc) para su análisis en la siguiente sección.**

### 📌Entregable: Competencia en CodaLab

<center>
<img src="https://i.giphy.com/NomCzPIGoXs3EIq77v.webp" width="300" height="300">
</center>

Como parte de esta entrega final, el equipo debe utilizar su pipeline entrenado para **generar predicciones en los nuevos conjuntos de datos mencionados anteriormente**.

Estas predicciones deben:

- Ser generadas directamente desde el *pipeline* previamente desarrollado.  
- Guardarse en archivos `.csv` siguiendo el formato requerido.  
- Subirse a la plataforma **CodaLab**, donde se realizará la evaluación final.


**IMPORTANTE:** Subir los resultados a tiempo en las tres fechas es **<u>OBLIGATORIO</u>** para la evaluación del desempeño final del equipo. **Por cada fecha en la que no se suban predicciones, se aplicará un descuento de 0.75 puntos (75 décimas) sobre la nota de la Entrega 3.**

### 🎁 Bonus [0.5 puntos]

Como incentivo adicional, se premiará a los **3 equipos con mejor performance en la métrica F1** con **0.5 puntos de puntaje adicional** sobre la nota de la Entrega 3.

## 📌 Informe: Performance del modelo [6.0 puntos]

En esta sección se espera que el equipo analice de manera retrospectiva el performance de su modelo y saque conclusiones en función de sus resultados. 

Para esto, tendrán que incurrir en 3 tipos de análisis:

- **Individual:** Deben evaluar el performance de su modelo tomando cada semana de forma aislada.

- **Comparativo:** Deben evaluar el performance de su modelo *a través* de las semanas, comparando el desempeño entre las semanas e identificando posibles tendencias.
    - Para esta parte se espera que generen gráficos de tendencia y tablas comparativas para apoyar su análisis.

- **Conclusiones y Aprendizajes:** Deben escribir sus principales conclusiones y aprendizajes de este proyecto. 

A lo largo de esta sección, se espera que respondan preguntas como:

1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 
2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?
3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron?  
4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?
4. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  
5. ¿Qué hiperparámetro fue el más importante para su modelo? 
5. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?
6. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?
7. ¿Qué limitaciones detectaron en su modelo o en los datos?  

**IMPORTANTE: Se espera que en sus respuestas hagan referencia a los artefactos (métricas, hiperparámetros, gráficos de interpretabilidad, etc) que su pipeline genera.**

<center>
<img src="https://media1.tenor.com/m/jyQ3SPT1htEAAAAd/i-love-this-performance-even-more-simon-cowell.gif" width="300" height="300">
</center>

Los archvios batch son de la forma Usuario x producto, y solo tienen los que cumplen que la proba de compra es mayor al ubral usado en esa fase, 50% fase 1, y 25% en las siguientes.

El archvio concatenar fue para concatenar la data, y el trabajo, fue para ir generando los csv. El informe esta con el que mandaron al correo las aux

### 📌 Análisis individual [3.0 puntos]

In [1]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

clientes_df = pd.read_parquet('clientes.parquet')
productos_df = pd.read_parquet('productos.parquet')

universe_size = len(clientes_df) * len(productos_df)

def get_batch_performance(batch_id):
    # Cargar datos reales
    data = pd.read_parquet(f"data/batch_t{batch_id}.parquet")
    data = data.groupby(['customer_id', 'product_id']).size().reset_index()
    true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))

    # Cargar predicciones
    preds = pd.read_csv(f"predictions/batch_t{batch_id}.csv",
                        names=['customer_id', 'product_id'])
    pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))

    # TP = predicho y real
    tp = len(pred_pairs & true_pairs)
    # FP = predicho pero no real
    fp = len(pred_pairs - true_pairs)
    # FN = real pero no predicho
    fn = len(true_pairs - pred_pairs)
    # TN = universo total - todo lo demás
    tn = universe_size - tp - fp - fn

    y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
    y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

    metrics = {
        "accuracy": accuracy_score(y_true, y_pred),
        "precision": precision_score(y_true, y_pred),
        "recall": recall_score(y_true, y_pred),
        "f1": f1_score(y_true, y_pred)
    }

    return metrics

FileNotFoundError: [Errno 2] No such file or directory: 'clientes.parquet'

#### 📌 Batch 1 (01/01/25 - 05/01/25) [0.75 puntos]

In [ ]:
get_batch_performance(1)

{'accuracy': 0.9978503431902482,
 'precision': 0.4936416184971098,
 'recall': 0.13082107843137256,
 'f1': 0.20682974085735045}

In [ ]:
# aquí el código 
# concatenar el batch 1 a los datos historicos para reentrenar y obtener modelo
get_batch_performance(1)

top_X = 5  # Número de recomendaciones por cliente
#obytenemos las top x predicciones para cada cliente
df_results_batch1 = pd.read_parquet("predictions/predictions_V2.parquet")
top_predictions = df_results_batch1.groupby('customer_id', group_keys=False).apply(
    lambda x: x.nlargest(top_X, 'probabilidad_compras')
)



/var/folders/dv/2lkcl0fn3wjgx7zy0djq3pkh0000gn/T/ipykernel_83727/4273468843.py:7: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_predictions = df_results_batch1.groupby('customer_id', group_keys=False).apply(


In [ ]:
#Sacamos metriocas para este cliente
data = pd.read_parquet(f"data/batch_t1.parquet")
data = data.groupby(['customer_id', 'product_id']).size().reset_index()
true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))

# Cargar predicciones
preds = top_predictions
pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))

# TP = predicho y real
tp = len(pred_pairs & true_pairs)
# FP = predicho pero no real
fp = len(pred_pairs - true_pairs)
# FN = real pero no predicho
fn = len(true_pairs - pred_pairs)
# TN = universo total - todo lo demás
tn = universe_size - tp - fp - fn

y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

metrics = {
    "accuracy": accuracy_score(y_true, y_pred),
    "precision": precision_score(y_true, y_pred),
    "recall": recall_score(y_true, y_pred),
    "f1": f1_score(y_true, y_pred)
}
metrics

{'accuracy': 0.9948145683062477,
 'precision': 0.18885906040268458,
 'recall': 0.43106617647058826,
 'f1': 0.26264700392010454}

En la primera semana, el modelo enfrentó un escenario en el que se tuvo que adaptar a los datos iniciales. Tras aplicar el filtro de que limita a los 5 productos mas probables, se obtuvo una Precisión de 18.9%, un Recall de 43.1% y un F1-Score de 0.26.

La precisión por cliente es la más baja de todos los batches, la mayoría de las recomendaciones no conlleva a compras, lo cual es esperable considerando que no cuenta con historia de una semana anterior. 

Al limitar los productos a 5 recomendaciones, se fuerza la aparición de productos relevantes, llevando a un recall de 43%, pero con una tasa de falsos positivos elevada. Es decir el modelo esta generalizando demasiado. Es muy buena practica limitarlo a los 5 mejores productos, ya que acota la cantidad opciones con lo que tienen que lidiar las metricas. Se puede ver una mejora de 6 puntos al poner el filtro.

El bajo F1 0.26 es indicativo de que no es un modelo aplicable y suficientemente aplicable para ejercerse en el mercado.

#### 📌 Batch 2 (06/01/25 - 12/01/25) [0.75 puntos]

In [ ]:
get_batch_performance(2)

{'accuracy': 0.995741382173536,
 'precision': 0.336094674556213,
 'recall': 0.2782676856750931,
 'f1': 0.30445969125214406}

In [ ]:
top_X = 5 # Número de recomendaciones por cliente
df_results_batch2 = pd.read_parquet("predictions/predictions_fase2.parquet")
top_predictions = df_results_batch2.groupby('customer_id', group_keys=False).apply(
    lambda x: x.nlargest(top_X, 'probabilidad_de_compra')
)

data = pd.read_parquet(f"data/batch_t2.parquet")
data = data.groupby(['customer_id', 'product_id']).size().reset_index()
true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))

# Cargar predicciones
preds = top_predictions
pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))

# TP = predicho y real
tp = len(pred_pairs & true_pairs)
# FP = predicho pero no real
fp = len(pred_pairs - true_pairs)
# FN = real pero no predicho
fn = len(true_pairs - pred_pairs)
# TN = universo total - todo lo demás
tn = universe_size - tp - fp - fn

y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

metrics = {
    "accuracy": accuracy_score(y_true, y_pred),
    "precision": precision_score(y_true, y_pred),
    "recall": recall_score(y_true, y_pred),
    "f1": f1_score(y_true, y_pred)
}
metrics

/var/folders/dv/2lkcl0fn3wjgx7zy0djq3pkh0000gn/T/ipykernel_83727/960008257.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_predictions = df_results_batch2.groupby('customer_id', group_keys=False).apply(


{'accuracy': 0.9944062976083345,
 'precision': 0.270690811535882,
 'recall': 0.395453654712914,
 'f1': 0.32138875617136486}

En la segunda semana, tras  reentrenar, se puede ver un salto significativo en los valores. Se obtuvo una Precisión de 27.1%, un Recall de 39.5% y un F1-Score de 0.32. El recall cae levemente, como 4 puntos, pero la precisión aumenta más de 8%.

La precisión por cliente mejora significativamente, llevando a muchos usuarios a compras. Aunque el desbalance persiste, es decir, la relación entre productos recomendados y compras se hace más correcto.

Con el F1 de 0.32, demuesta que el modelo ha aprendido a discriminar mejor. Sacrificar un poco de recall, que seria como la cobertura, ayudo en limpiar una gran porción de falsos positivos.

También se tiene que mensionar que aunque se limite a 5 productos ya no ve una mejora substancial comparado con el modelo que cuenta con todos los resultados.

En conclusión si se vio una mejora notable comparada con el batch anterior, y si fuera aplicarse se desempeñaria mucho mejor que el anterior.

#### 📌 Batch 03 (13/01/25 - 19/01/25) [0.75 puntos]

In [ ]:
get_batch_performance(3)

{'accuracy': 0.995975711175393,
 'precision': 0.3422083007047768,
 'recall': 0.26635514018691586,
 'f1': 0.29955443847823604}

In [ ]:
top_X = 5  # Número de recomendaciones por cliente
df_results_batch3 = pd.read_parquet("predictions/predictions_fase_3.parquet")
top_predictions = df_results_batch3.groupby('customer_id', group_keys=False).apply(
    lambda x: x.nlargest(top_X, 'probabilidad_de_compra')
)

data = pd.read_parquet(f"data/batch_t3.parquet")
data = data.groupby(['customer_id', 'product_id']).size().reset_index()
true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))

# Cargar predicciones
preds = top_predictions
pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))

# TP = predicho y real
tp = len(pred_pairs & true_pairs)
# FP = predicho pero no real
fp = len(pred_pairs - true_pairs)
# FN = real pero no predicho
fn = len(true_pairs - pred_pairs)
# TN = universo total - todo lo demás
tn = universe_size - tp - fp - fn

y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

metrics = {
    "accuracy": accuracy_score(y_true, y_pred),
    "precision": precision_score(y_true, y_pred),
    "recall": recall_score(y_true, y_pred),
    "f1": f1_score(y_true, y_pred)
}
metrics

/var/folders/dv/2lkcl0fn3wjgx7zy0djq3pkh0000gn/T/ipykernel_83727/3423726170.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_predictions = df_results_batch3.groupby('customer_id', group_keys=False).apply(


{'accuracy': 0.9945211647661075,
 'precision': 0.2712090848363393,
 'recall': 0.4124339699309224,
 'f1': 0.3272346256145724}

En la tercera semana, el modelo alcanza su mejor desempeño global. Se registraron métricas de Precisión 27.1%, Recall 41.2% y F1-Score 0.33. El recall mejora pero la precision se mantiene en su misma ubicación.

La precisión, la calidad, sigue siendo de 1 en cada 4 productos sugeridos de la selección de los top 5. El F1 llega a 0.327 que es casi idéntico al del Batch 2, pero con mejor recall, es decir tiene una mejor cobertura de productos.

En este caso también se ve una leve mejora con el filtro de top 5, mejorando el f1 en 2 puntos

Operativamente, el modelo sigue generando muchas recomendaciones, pero cabe recalcar que el F1 se ha optimizado. El filtro de los tops esta funcionando bien.

#### 📌 Batch 04 (20/01/25 - 26/01/25) [0.75 puntos]

In [ ]:
# aquí el código 
# concatenar el batch 4 a los datos historicos para reentrenar y obtener modelo
get_batch_performance(4)

{'accuracy': 0.9961358688125165,
 'precision': 0.3590168461750898,
 'recall': 0.2671598849157419,
 'f1': 0.3063508895958525}

In [ ]:
top_X = 5  # Número de recomendaciones por cliente
df_results_batch4 = pd.read_parquet("predictions/predictions_fase_4.parquet")
top_predictions = df_results_batch4.groupby('customer_id', group_keys=False).apply(
    lambda x: x.nlargest(top_X, 'probabilidad_de_compra')
)

data = pd.read_parquet(f"data/batch_t4.parquet")
data = data.groupby(['customer_id', 'product_id']).size().reset_index()
true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))

# Cargar predicciones
preds = top_predictions
pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))

# TP = predicho y real
tp = len(pred_pairs & true_pairs)
# FP = predicho pero no real
fp = len(pred_pairs - true_pairs)
# FN = real pero no predicho
fn = len(true_pairs - pred_pairs)
# TN = universo total - todo lo demás
tn = universe_size - tp - fp - fn

y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

metrics = {
    "accuracy": accuracy_score(y_true, y_pred),
    "precision": precision_score(y_true, y_pred),
    "recall": recall_score(y_true, y_pred),
    "f1": f1_score(y_true, y_pred)
}
metrics

/var/folders/dv/2lkcl0fn3wjgx7zy0djq3pkh0000gn/T/ipykernel_83727/108535482.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_predictions = df_results_batch4.groupby('customer_id', group_keys=False).apply(


{'accuracy': 0.9945264158361771,
 'precision': 0.26892880904856953,
 'recall': 0.41533086724208795,
 'f1': 0.3264679751231726}

En la cuarta semana, el sistema muestra estabilidad, es decir, ningun cambio. Se obtuvo una Precisión de 26.9%, un Recall de 41.5% y un F1-Score de 0.33, las variaciones, minimas

La precisión de 27%, se mantiene igual y el desbalance entre recomendaciones y reales ya es esperable, de un punto de vista de una empresa puede ser algo bueno ya que permite preparse.

El comportamiento identico al de la semana anterior confirma que el modelo no tiene un drift significativo.

Operativamente, el modelo es confiable. Pero se estanca con los resultados, resultados que se tienen que analizar para estimar el costo de oportunidad por cada producto. Si se quiere mejorar, no basta con reentrenar, sino que tambien va a necesitar nuevas variables.

### 📌 Análisis comparativo [2.0 puntos]

In [ ]:
# Aquí se espera que puedan generar gráficos para ver como fue la evolución de las métricas a lo largo de las semanas/batches 
# En caso de que corrieron todos los modelos con Optuna y SHAP pueden adjuntar los gráficos como
# - Optimization History Plot
# - Importance Plot
# - SHAP Summary Plot

**Conclusiones**

El sistema muestra un comportamiento dual. Por un lado, alcanza una cobertura.....

1. Las metricas se separan en dos fases, el del batch 1, en que tuvo que modelar sin historia, lo cual llevo a un peor desempeño, y el resto de los batch, en el que ya estaba mas estable, llevan las metricas a un mejor puntaje. Se vio una mejora de F1 de 0.26 a 0.32. También se vio un mucho mejor recall. Es decir que con una semana de información ya mejora el modelo

También, se puede ver una mejora clara una vez que se implementa el filtro de los tops 5, en el que se debe destacar en especial en el primer batch, que subio de 0.2 a 0.26. Y que despues se estabilizo con con el los batches 2-3-4, manteniendose en 0.3 para el f1

Es importante notar que el recall, que es la cobertura son las la gamma de recommendaciones, que son productos que puede llegar a comprar. Algo que tratamos de priorizar, considerando que nos interesa la recomendacion al cliente, sin embargo, la precicion fue mas baja, lo que llevo a que estuvieramos presentando productos que no llevarian a compras, a lo que lleva a explicar que tiene mala calidad las recomendaciones.

2. El peor desempeño sucedio en el Batch 1. La razón principal fue la falta de lags. Al ser la primera semana, el modelo no disponía de datos con el que comparar. Esto obligó al algoritmo a generalizar basándose en las variables que le entregamos nosotros o historicos muy antiguos, lo que redujo la precisión.



1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 

* Comparar comportamiento sin filtro vs con Top-N
* Analizar trade-off entre precisión/F1 y recall
* Identificar estabilidad del sistema y rol del umbral en el balance calidad-cobertura


2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?

* Identificar batch con peor rendimiento
* Explicar causas: falta de historial, calibración de probabilidades
* Mencionar mejora progresiva con acumulación de datos

3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron? 

* Describir metodología de detección (tests estadísticos para variables numéricas y categóricas, KS chi-cuadrado)
* Explicar sistema de scoring y umbral de activación
* Analizar si el drift detectado fue real o producto de sensibilidad del detector
* Discutir posibles causas: historial limitado, umbral muy bajo

4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?

* Mencionar frecuencia de reentrenamiento y razón (activación por drift)
* Analizar impacto diferenciado: mejora en cobertura vs estancamiento en precisión
* Explicar por qué mejoró una métrica pero no la otra (relación con umbral de decisión)
* Conclusión sobre trade-off cobertura-relevancia

5. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  

* Identificar decisión con mayor impacto positivo (feature engineering)
* Especificar tipos de variables creadas (recencia, agregaciones temporales)
* Identificar decisión con mayor impacto negativo (política de umbral)
* Explicar cómo afectó cada decisión al desempeño final

6. ¿Qué hiperparámetro fue el más importante para su modelo?

* Mencionar herramienta de optimización utilizada
* Listar hiperparámetros más influyentes identificados
* Explicar rol de cada uno: manejo de desbalance, regularización, captura de interacciones
* Conectar ajuste de hiperparámetros con métrica final alcanzada


7. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?

Los feature importances de los cuatro modelos XGBoost muestran que el modelo depende casi por completo de la recencia de compra. Las variables más influyentes fueron:

 - num__items_roll4_mean 

- num__items_last_week

- cat__sub_category_GASEOSAS

- cat__week_1

- cat__brand_Brand 24

Las más consistentes en todos los batches fueron:

- num__items_roll4_mean
- num__items_last_week
- cat__week_1 
- cat__brand_Brand 24


- num__items_roll4_mean fue la variable dominante en todos los batches (35–48% del gain), aunque bajó ligeramente del Batch 1 al 4.
- num__items_last_week también se mantuvo estable en el Top 5, con una leve caída en su aporte.
Entre las categóricas, cat__sub_category_GASEOSAS ganó relevancia en los batches intermedios, especialmente en el Batch 3.

En conjunto, las variables de recencia siguieron siendo las más predictivas, pero su descenso marginal indica que el modelo empezó a incorporar señales adicionales como categorías, marcas y patrón semanal.


8. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?

* Acá ver patrones de comportamiento como importancia de compras recientes, estacionalidad
* Diferencias por categoría/marca (segmentacion)
* Comentar performance, balance cobertura-relevancia, volumen de recomendaciones
* Consistencia de variables clave en el tiempo
* Algunas recomendacoinse estrategias para mejorar eficiencia (filtros, límites, personalización)

9. ¿Qué limitaciones detectaron en su modelo o en los datos?

* Por ej el volumen excesivo de recomendaciones y su impacto, proporción de positivos y consecuencias del submuestreo, problemas con la estrategia de split temporal y optimismo en métricas, sugerencias de mejoras en nuestro caso un esquema de evaluación más realista

### 📌 Conclusiones y Aprendizajes [1.0 puntos]

Este proyecto permitió ....

## 📁 Estructura de entrega

Se espera que su entrega cuente con la siguiente estructura de carpetas:

```bash
entrega_3/
│
├── informe.ipynb # Jupyter notebook con sus resultados
│
├── predictions/ # Carpeta con las predicciones para cada batch de datos
│
└── data/ # Carpeta con los datos realizados (lo que realmente ocurrió)
```

Mucho éxito!

<center>
<img src="https://gifdb.com/images/high/may-the-force-be-with-you-anakin-skywalker-n2g5o0pm4h6iylsx.gif" width="400" height="300">
